### PR TITLE
Create priority_queue.py

### PR DESCRIPTION
**Pull request by Mayureshd-18**
*Please read the [contributing guidelines](https://github.com/huangsam/ultimate-python/blob/master/CONTRIBUTING.md) before submitting a pull request.*

---

**Describe the change**
Adding the priority queue data structure in python which is mainly implemented using the heapq module in python

**Additional context**
Added the explanation and the same output below the code.

### CODE REVIEW

**Comment by Mayureshd-18** [comment]
Didnt write the name of the file as heapq.py as it may cause circular imports error

**Comment by Mayureshd-18** [comment]
@huangsam Made the requested changes. Please review.

File: ultimatepython/data_structures/priority_queue.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    """
+    Demonstrates the usage of the heapq module in Python.
+    """
+
+    # Step 1: Creating a Min Heap
+    # Initialize a list with some elements
+    heap = [3, 1, 4, 1, 5, 9, 2]
+
+    # Convert the list into a min heap using heapq.heapify()
+    heapq.heapify(heap)
```
Comment (huangsam) [inline ultimatepython/data_structures/priority_queue.py:14]: Should we share a video or informational about what `heapify` does under the hood?

File: ultimatepython/data_structures/priority_queue.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+    # Assert that the heap is in min heap order
+    assert heap == [1, 1, 2, 3, 5, 9, 4]
```
Comment (huangsam) [inline ultimatepython/data_structures/priority_queue.py:17]: Should we share what it means for a Python list to be in min heap order?

File: ultimatepython/data_structures/priority_queue.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+    # Step 2: Pushing onto the Heap
+    # Add a new element to the heap using heapq.heappush()
```
Comment (huangsam) [inline ultimatepython/data_structures/priority_queue.py:20]: Again, should we be sharing what happens under the hood for `heappush`?

File: ultimatepython/data_structures/priority_queue.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+# The function maintains the heap property, ensuring the smallest element remains at the root.
+# Popping from the Heap (heappop):
+
+# Remove and retrieve the smallest element from the heap using heapq.heappop().
+# The heap is automatically adjusted to maintain the heap property.
+# Heap Replacement (heapreplace):
+
+# Replace the smallest element with a new value using heapq.heapreplace().
+# This operation efficiently maintains the heap property by removing and adding elements.
+# By following these steps, you can effectively create, modify, and work with min heaps using the heapq module in Python.
```
Comment (huangsam) [inline ultimatepython/data_structures/priority_queue.py:67]: Perhaps this content should be put at the very top of the module in the module's docstring

File: ultimatepython/data_structures/priority_queue.py [hunk 5]
```diff
@@ -0,0 +1,67 @@
+import heapq
+
+
+def main():
+    """
+    Demonstrates the usage of the heapq module in Python.
+    """
```
Comment (huangsam) [inline ultimatepython/data_structures/priority_queue.py:7]: This should be put at the top of the module, along with the content at the bottom of the file

**Review by huangsam — CHANGES_REQUESTED** [review]
In general a good resource for this is:
https://docs.python.org/3/library/heapq.html#theory

Please make it such that folks understand what's happening under the hood. Otherwise, people would just prefer using the `list.sort` or `sorted` method. Lots of Python devs want sorting to "just work" - so I don't think sharing just the steps for Heap usage is enough

**Comment by huangsam** [comment]
This PR has been around for too long. Let's close it for now. If you would like to make a new PR again, feel free to submit another one. Thanks again for the submission

**Pull request closed without merging**